packages/mobile/src/screens/chat-screen/ChatScreen.tsx

```diff
@@ -92,22 +101,81 @@ export const ChatScreen = () => {
 
   const { params } = useRoute<'Chat'>()
   const { chatId } = params
-  const url = `/chat/${encodeUrlName(chatId)}`
-  const [iconOpacity, setIconOpacity] = useState(0.5)
-  const [text, setText] = useState('')
+  const url = `/chat/${encodeUrlName(chatId ?? '')}`
+  const [iconOpacity, setIconOpacity] = useState(ICON_BLUR)
+  const [inputMessage, setInputMessage] = useState('')
+  const [earliestUnreadIndex, setEarliestUnreadIndex] = useState<
+    number | undefined
+  >()
 
+  const userId = encodeHashId(useSelector(getUserId))
+  const chat = useSelector((state) => getChat(state, chatId ?? ''))
+  const [otherUser] = useSelector((state) => getOtherChatUsers(state, chatId))
   const chatMessages = useSelector((state) =>
     getChatMessages(state, chatId ?? '')
   )
   const status = useSelector((state) =>
     getChatMessagesStatus(state, chatId ?? '')
   )
+  const summary = useSelector((state) =>
+    getChatMessagesSummary(state, chatId ?? '')
+  )
+  const flatListRef = useRef<RNFlatList>(null)
+
+  // A ref so that the unread separator doesn't disappear immediately when the chat is marked as read
+  // Using a ref instead of state here to prevent unwanted flickers.
+  // The chat/chatId selectors will trigger the rerenders necessary.
+  const chatFrozenRef = useRef(chat)
 
   useEffect(() => {
-    dispatch(fetchMoreMessages({ chatId }))
-  }, [dispatch, chatId])
+    if (chatId && status === Status.IDLE) {
+      // Initial fetch
+      dispatch(fetchMoreMessages({ chatId }))
+    }
+  }, [dispatch, chatId, status, summary])
 
-  const [otherUser] = useSelector((state) => getOtherChatUsers(state, chatId))
+  useEffect(() => {
+    // Update chatFrozenRef when entering a new chat screen
+    if (chat && chatId !== chatFrozenRef.current?.chat_id) {
```

Comment:
@rickyrombo I moved the ref setting up bc I think we actually want to make sure we don't call `findIndex` with previous chat params. Seems to work on testing, does this make sense?

Reply:
Yeah this is fine, especially as you called out we want the index to change when the user sends/receives a new message so using the effect/state etc is all moot... sorry for the wrong advice there